src/jit/codegenxarch.cpp

```diff
@@ -2470,6 +2470,18 @@ CodeGen::genCodeForTreeNode(GenTreePtr treeNode)
     case GT_JTRUE:
         {
             GenTree *cmp = treeNode->gtOp.gtOp1;
+
+#if !defined(_TARGET_64BIT_)
+            GenTreePtr op1 = cmp->gtGetOp1();
+            bool isSigned = ((cmp->gtFlags & GTF_UNSIGNED) == 0);
+
+            // For long signed compares that are not GT_NE or GT_EQ, we emit the logic in genCompareLong,
+            // and do not need to do anything here.
+            if (varTypeIsLong(op1->TypeGet()) && isSigned && (cmp->gtOper != GT_NE && cmp->gtOper != GT_EQ))
+            {
+                break;
+            }
+#endif
             assert(cmp->OperIsCompare());
```

Comment:
It seems like doing all the work in genCompareLong() leads to another silent bad codegen bug. I am working on a fix that moves the work for the case where the result isn't realized in a register to the JTRUE case. It will split the logic, but fixes the silent bad codegen.